File: 14-spring-security/backend/src/main/java/ru/sber/services/CartService.java
package ru.sber.services;

import ru.sber.entities.Product;

import java.util.List;

/**
 * Сервис для взаимодействия с корзиной пользователя
 */
public interface CartService {
    /**
     * Добавление товара в корзину
     *
     * @param userId    Уникальный идентификатор пользователя
     * @param productId Уникальный идентификатор товара
     * @return Возвращает статус добавления товара в корзину
     */
    boolean addToCart(long userId, long productId);

    /**
     * Изменение количества товара в корзине
     *
     * @param userId    Уникальный идентификатор пользователя
     * @param productId Уникальный идентификатор товара
     * @param amount    Количество добавляемого товара
     * @return Возвращает статус обновления количества товара в корзине
     */
    boolean updateProductAmount(long userId, long productId, int amount);

    /**
     * Удаление товара из корзины
     *
     * @param userId    Уникальный идентификатор пользователя
     * @param productId Уникальный идентификатор товара
     * @return Возвращает статус удаления товара из корзины
     */
    boolean deleteProduct(long userId, long productId);

    /**
     * Полностью очищает корзину пользователя
     *
     * @param userId Уникальный идентификатор пользователя
     */
    void clearCart(long userId);

    /**
     * Выдает список товаров в корзине пользователя
     *
     * @param userId Уникальный идентификатор пользователя
     * @return Возвращает список товаров
     */
    List<Product> getListOfProductsInCart(long userId);

    /**
     * Подсчитывает количество товаров в корзине пользователя
     *
     * @param userId Уникальный идентификатор пользователя
     * @return Возвращает количество товаров
     */
    int countProductsInCart(long userId);

}

File: 14-spring-security/frontend/src/services/cartService.js
import axios from "axios";
import authHeader from "./auth-header";
import {set} from "../slices/userSlice";
import {setCart} from "../slices/cartSlice";

const API_URL = "http://localhost:8081/cart";

const getCart = (userId, dispatch) => {

    return axios.get(API_URL + `/${userId}`,{headers: authHeader()}).then(
        (response) => {
            console.log(response.data)
            dispatch(setCart(response.data))
        },
        (error) => {
            const _content = (error.response && error.response.data) ||
                error.message ||
                error.toString();

            console.error(_content)

            dispatch(set([]));
        });
}

const addToCart = (userId, productId, dispatch) => {

    console.log(`${API_URL}/${userId}/product/${productId}`, {quantity: 1}, {headers: authHeader()})

    return axios.post(`${API_URL}/${userId}/product/${productId}`, {quantity: 1}, {headers: authHeader()}).then(
        () => {
            getCart(userId, dispatch)
        },
        (error) => {
            const _content = (error.response && error.response.data) ||
                error.message ||
                error.toString();

            console.error(_content)
        });
};

const updateAmount = (userId, productId, amount, dispatch) => {
    return axios.put(`${API_URL}/${userId}/product/${productId}`, amount, {headers: authHeader()}).then(
        () => {
            getCart(userId, dispatch)
        },
        (error) => {
            const _content = (error.response && error.response.data) ||
                error.message ||
                error.toString();

            console.error(_content)
        });
};

const deleteFromCart = (userId, productId, dispatch) => {

    return axios.delete(`${API_URL}/${userId}/product/${productId}`, {headers: authHeader()}).then(
        () => {
            getCart(userId, dispatch)
        },
        (error) => {
            const _content = (error.response && error.response.data) ||
                error.message ||
                error.toString();

            console.error(_content)
        });
};

const clearCart = (state) => {
    state.cartItems = [];
}

const cartService = {
    getCart,
    addToCart,
    deleteFromCart,
    updateAmount,
    clearCart
};

export default cartService